APEXMALHAR-1972 Added support for Expression Evaluator

chinmaykolhatkar opened the pull request:
Added support for Expression Evaluator for quasi-Java Expression Language.

Design details are all given in Jira.

Commit:
APEXMALHAR-1972 Added support for Expression Evaluator for quasi-Java Expression Language.

shubham-pathak22 commented on library/src/main/java/com/datatorrent/lib/expressions/ConversionUtils.java:
```diff
@@ -0,0 +1,204 @@
+    return new Short(a).intValue();
+  }
+
+  public static int toInt(String a)
```
can we add java docs ? NumberUtils returns 0 for null , blank or incorrect string.

shubham-pathak22 commented on library/src/main/java/com/datatorrent/lib/expressions/ExpressionEvaluator.java:
```diff
@@ -0,0 +1,440 @@
+ * 2. Return type of of expression evaluation.
+ * 3. Expression to be evaluated. This is a standard java expression except for referencing the variable inside object JEL syntax needs to be used i.e. ${objectPlaceHolder.varName}
+ *
+ * Expression cab be defined as per JSL 7 syntax:
```
can

shubham-pathak22 commented on library/src/main/java/com/datatorrent/lib/expressions/ExpressionEvaluator.java:
```diff
@@ -0,0 +1,440 @@
+    }
+
+    Pattern entry = Pattern.compile(variablePlaceholderPattern);
+    Matcher matcher = entry.matcher(expression);
```
Should we consider throwing runtime exception if expression is null / blank ?

chinmaykolhatkar replied:
Considering this will be run in some operator, when the expression variable is encountered, it'll anyway throw exception. 
Do you see any advantage in adding this check before that?

Commit:
Added comments to util class and methods.
Taken care of few review comments.

Commit:
Changing the inteface of ExpressionEvaluator.
This gives a way to override expression parsing and expression evaluation logic.
